wlcreate/Start-Asian-Love
# Home news branch

## Description



Related Issue: Closes #3 



- **On Saturday, April 24th, 2021, a news page was set up to fetch articles from the [News API](https://newsapi.org/).** 
    - In the root directory, a news folder with a file called [slug].js was created.
    - The news code comes largely from the Next.js news API [tutorial](https://youtu.be/xtItzwYG6oQ) by Zackery Wilson.
    - The fetch request has a query parameter for "stopasianhate" and "asians."
    
    - A few domains were excluded due to lack of relevancy or lack of content/images.
    - News articles are displayed as a grid, with 9 articles displayed across 10 pages.
    - Currently, the article cards expand to show the first 3 lines of article content upon hover. This effect is to minimize visual clutter (and the effect may be used in the Support and Wins pages), but the cards can be reverted back to its initial stationary form.

- **In the early evening of Sunday, April 25th, the news page failed to compile properly.**
    -  Reason: 'undefined' variables cannot be serialized as JSON, in regards to **getServerSideProps**.
    - I suspect the undefined values may come from empty "publishedAt" fields in some news articles.
    - Github discussion about this error: https://github.com/vercel/next.js/discussions/11209
    - Developers want this issue to be reopened by Next.js/Vercel, as it is a persistent problem that should be accounted for by JSON.parse(JSON.stringify(props)).
    
    - This **getServerSideProps** error came up when I made my own version of a Next.JS news site earlier in April, and the error also appeared during the weekend, but not on the weekdays.
    - On Monday morning, April 26th, the news page was still failing to load, so I emailed Zackery Wilson, who replied that he will investigate the issue this week and respond as soon as his schedule opens up.
    - On Monday evening, the news page inexplicably started working again, albeit having no relevant code changed (same pattern as my prior Next.js news project).
    - On Wednesday morning, April 28th, the news page again failed to load, but I realized that the `.env` file was absent. I suspect that committing an `.env` file on a sub branch while it is included in the `.gitignore` file makes the file disappear entirely, unless it is on the main branch. 
    - The news page is now working again with the addition of the `.env` file and the API key.
  
- **In the Home page, some initial styling has been applied (all subject to change!).**
    - Overall, the text has been made larger, and the left-hand side logo in the header also links back to the home page.
    - The video player is now nested within two divs in order to reduce size.
    - I experimented with putting the video player in the middle of the text, but it can be positioned anywhere.
    - I attempted to show a "news ticker"/ first 3 or 4 articles from the News API fetch request under the video player in the home page, but I haven't found a solution to implement it yet.
    
    - The images and text in the carousel were enlarged and now cover the full width, and the arrows are temporarily commented out.
    - Some initial media queries have been applied in the header.module.css, but values need to be refined for consistency across all pages.
    - Styling is still a work in progress, but the site is shaping up!
    - **Styling can definitely be rearranged and cleaned up into Global CSS, it's a bit of a hot mess right now** 😂

## Resources: 
- Next.js News Site Tutorial: https://youtu.be/xtItzwYG6oQ
- Github discussion about getServerSideProps: https://github.com/vercel/next.js/discussions/11209
- The pull request that developers want Next.js to have re-opened: https://github.com/vercel/next.js/pull/17209
- Truncating lines of text, resource 1: https://css-tricks.com/line-clampin/
- Truncating lines of text, resource 2: https://dropshado.ws/post/1015351370/webkit-line-clamp
- CSS zoom hover effects and overflow property: https://w3bits.com/css-image-hover-zoom/

## Acceptance Criteria

- Use one of the follow APIs to let the user see news related to anti-Asian hate, Asian racism, and the Stop Asian Hate movement
    - https://newscatcherapi.com/free-news-api
    - https://mediastack.com
- Resource
    - [News API Tutorial](https://youtu.be/xtItzwYG6oQ)
- Optional
    - Pagination
    - Filtering/Sorting capabilities based on date
    - Filtering by location
 - Rallies and Newest Updates
    - Provide most current information on rallies and news articles



## Self-check

Please check all that apply:



## Type of Changes



|     | Type                       |
| --- | -------------------------- |
|    | :bug: Bug fix              |
|  ✓  | :sparkles: New feature     |
|    | :hammer: Refactoring       |
|    | :100: Add tests            |
|    | :link: Update dependencies |
|    | :scroll: Docs              |

## Updates

### Before


![home-before](https://user-images.githubusercontent.com/61435324/116483497-33391e80-a855-11eb-9ec2-d1af4320428a.gif)
- **Home page before additional styling.**

### After


![home-after1](https://user-images.githubusercontent.com/61435324/116483534-40eea400-a855-11eb-8e77-012c90de3441.gif)
- **Home page with enlarged text and resizing of carousel slides.**


![home-after2](https://user-images.githubusercontent.com/61435324/116484673-89a75c80-a857-11eb-8c5c-b0e9f1395d36.gif)
- **The video player has been resized and temporarily put in the middle of the Home page text.**

- **The "check out the latest news" is completely optional and it's a placeholder for the possible implementation of a news ticker/news API feed.**


![news1](https://user-images.githubusercontent.com/61435324/116483654-7e533180-a855-11eb-89d0-0391f404b5be.gif)
- **For now, the news article cards expand upon hover, and the cards show 3 truncated lines of text. Upon click, the articles open in a new browser tab.**


![news2](https://user-images.githubusercontent.com/61435324/116483739-ac387600-a855-11eb-96d0-f4d6569e66b5.gif)
**News articles are displayed in a 3x3 grid, with a total of 10 news pages with 9 articles each.**


# 

## Testing Steps / QA Criteria


- Switch to the home-news branch.
- If `.env` file is missing, create the `.env` file in the main directory, and it should appear under the `styles` folder. API key is in Slack message.
- Click on the News Page to see if it is compiling successfully.
- Hover and click on the news article cards.
- Click the paginator at the bottom of the news page, and check all 10 pages.
- On the Home page, click on the dots below the carousel, and resize the browser to see how carousel slides respond.
- Check the video player in the middle of the home page and resize browser for appearance changes.
